core/src/main/java/forklift/consumer/MessageRunnable.java

```diff
@@ -41,52 +43,50 @@
     @Override
     public void run() {
         RunAsClassLoader.run(classLoader, () -> {
-            try {
-                try {
-                    // Validate the class.
-                    LifeCycleMonitors.call(ProcessStep.Validating, this);
-                    for (Method m : onValidate) {
-                        if (m.getReturnType() == List.class) {
-                            addError((List<String>)m.invoke(handler));
-                        } else if (m.getReturnType() == boolean.class) {
-                            error = error || !((boolean)m.invoke(handler));
-                        } else {
-                            addError("Return type of " + m.getReturnType() + " is not supported for OnValidate methods");
-                        }
-                    }
-
-                    // Run the message if there are no errors.
-                    if (error) {
-                        LifeCycleMonitors.call(ProcessStep.Invalid, this);
-                    } else {
-                        LifeCycleMonitors.call(ProcessStep.Processing, this);
-                        for (Method m : onMessage) {
-                            // Send the message to each handler.
-                            m.invoke(handler, new Object[] {});
-                        }
-                    }
-                } catch (Throwable e) {
-                    log.info("Error processing", e);
-                    if (e.getCause() != null)
-                        addError(e.getCause().getMessage());
-                    else
-                        addError(e.getMessage());
+            // { Validating }
+            runHooks(ProcessStep.Validating);
+            LifeCycleMonitors.call(ProcessStep.Validating, this);
+            for (Method m : onValidate) {
+                if (m.getReturnType() == List.class) {
+                    addError(runLoggingErrors(() -> (List<String>)m.invoke(handler)));
+                } else if (m.getReturnType() == boolean.class) {
+                    boolean valid = runLoggingErrors(() -> (boolean)m.invoke(handler));
+                    if (!valid)
+                        addError("Validator " + m.getName() + " returned false");
+                } else {
+                    addError("onValidate method " + m.getName() + " has wrong return type " + m.getReturnType());
+                }
+            }
+
+            if (errors.size() > 0) {
+                // { Invalid }
+                runHooks(ProcessStep.Invalid);
+                LifeCycleMonitors.call(ProcessStep.Invalid, this);
+            } else {
+                // { Processing }
+                runHooks(ProcessStep.Processing);
+                LifeCycleMonitors.call(ProcessStep.Processing, this);
+                for (Method m : onMessage) {
+                    runLoggingErrors(() -> m.invoke(handler));
```

Comment:
For some reason the try/catch of the Throwable has been removed. So in the case that the message handler throws an NPE the consumer thread will die, and then restart. Also the message would be lost because it was Ack'd without completely processing through the lifecycle.